mlflow/diviner.py

```diff
@@ -192,7 +222,60 @@ def save_model(
     _PythonEnv.current().to_yaml(str(path.joinpath(_PYTHON_ENV_FILE_NAME)))
 
 
-def load_model(model_uri, dst_path=None):
+def _save_model_fit_in_spark(diviner_model, path: str, **kwargs):
+    """
+    Saves a Diviner model that was fit in Spark by processing the model components separately.
+    The metadata and ancillary files to write (JSON and Pandas DataFrames) are written directly
+    to a fuse mount location, which the Spark DataFrame that contains the individual serialized
+    Diviner model objects is written by using the 'dbfs:' scheme path that Spark recognizes.
+    """
+    # Validate that the path is a relative path early in order to fail fast prior to attempting
+    # to write the (large) DataFrame to a tmp DFS path first and raise a path validation
+    # Exception within MLflow when attempting to copy the temporary write files from DFS to
+    # the file system path provided.
+    if not os.path.isabs(path):
+        raise MlflowException(
+            "The save path provided must be a run-relative path. "
+            f"The path submitted, '{path}' is an absolute path."
+        )
+
+    # Create a temporary DFS location to write the Spark DataFrame containing the models to.
+    tmp_path = generate_tmp_dfs_path(kwargs.get("dfs_tmpdir", MLFLOW_DFS_TMP.get()))
+
+    # Save the model Spark DataFrame to the temporary DFS location
+    diviner_model._save_model_df_to_path(tmp_path, **kwargs)
+
+    diviner_data_path = os.path.abspath(path)
+
+    tmp_fuse_path = dbfs_hdfs_uri_to_fuse_path(tmp_path)
+    shutil.move(src=tmp_fuse_path, dst=diviner_data_path)
+
+    # Save the model metadata to the path location
+    diviner_model._save_model_metadata_components_to_path(path=diviner_data_path)
+
+
+def _load_model_fit_in_spark(local_model_path: str, flavor_conf, **kwargs):
+    """
+    Loads a Diviner model that has been fit (and saved) in the Spark variant.
+    """
+    # NB: To load the model DataFrame (which is a Spark DataFrame), Spark requires that the file
+    # partitions are in DFS. In order to facilitate this, the model DataFrame (saved as parquet)
+    # will be copied to a temporary DFS location. The remaining files can be read directly from
+    # the local file system path, which is handled within the Diviner APIs.
+    import diviner
+
+    dfs_temp_directory = generate_tmp_dfs_path(kwargs.get("dfs_tmpdir", MLFLOW_DFS_TMP.get()))
+    dfs_fuse_directory = dbfs_hdfs_uri_to_fuse_path(dfs_temp_directory)
+    os.makedirs(dfs_fuse_directory)
+    shutil_copytree_without_file_permissions(src_dir=local_model_path, dst_dir=dfs_fuse_directory)
```

Comment:
Q: why we cannot directly call `shutil.copytree` ?

Reply:
No specific reason. I'll test it :)

Reply:
Due to the nested file structure at the root of the directory, I'd have to reimplement the directory walking in order to get this to work. `shutil_copytree_without_file_permissions` already has the logic that I need in it to not overly complicate the implementation in the Diviner flavor.